packages/dds/map/src/directory.ts

```diff
@@ -832,15 +841,63 @@ export class SharedDirectory extends SharedObject<ISharedDirectoryEvents> implem
     }
 }
 
+interface IKeyEditLocalOpMetadata {
+    type: "edit";
+    pendingMessageId: number;
+    // "unknown" indicates we don't know the prior value because it is a stashed op so it cannot be rolled back
+    previousValue: ILocalValue | undefined | "unknown";
+}
+
+interface IClearLocalOpMetadata {
+    type: "clear";
+    pendingMessageId: number;
+    previousStorage: Map<string, ILocalValue> | "unknown";
+}
+
+interface ICreateSubDirLocalOpMetadata {
+    type: "createSubDir";
+    pendingMessageId: number;
+    previouslyExisted: boolean | "unknown";
+}
+
+interface IDeleteSubDirLocalOpMetadata {
+    type: "deleteSubDir";
+    pendingMessageId: number;
+    subDirectory: SubDirectory | undefined | "unknown";
+}
+
+type SubDirLocalOpMetadata = ICreateSubDirLocalOpMetadata | IDeleteSubDirLocalOpMetadata;
+type DirectoryLocalOpMetadata = IClearLocalOpMetadata | IKeyEditLocalOpMetadata | SubDirLocalOpMetadata;
+
+function isKeyEditLocalOpMetadata(metadata: any): metadata is IKeyEditLocalOpMetadata {
+    return metadata !== undefined && typeof metadata.pendingMessageId === "number" && metadata.type === "edit";
+}
+
+function isClearLocalOpMetadata(metadata: any): metadata is IClearLocalOpMetadata {
+    return metadata !== undefined && metadata.type === "clear" && typeof metadata.pendingMessageId === "number";
+}
+
+function isSubDirLocalOpMetadata(metadata: any): metadata is SubDirLocalOpMetadata {
+    return metadata !== undefined &&
+        ((metadata.type === "createSubDir" && typeof metadata.previouslyExisted !== undefined) ||
+         (metadata.type === "deleteSubDir" && typeof metadata.subDirectory !== undefined));
+}
+
+function isDirectoryLocalOpMetadata(metadata: any): metadata is DirectoryLocalOpMetadata {
+    return metadata !== undefined && typeof metadata.pendingMessageId === "number" &&
+        (metadata.type === "edit" || metadata.type === "clear" ||
+        metadata.type === "createSubDir" || metadata.type === "deleteSubDir");
+}
+
 /**
  * Node of the directory tree.
  * @sealed
  */
 class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirectory {
     /**
-     * Tells if the sub directory is disposed or not.
+     * Tells if the sub directory is deleted or not.
      */
-    private _disposed = false;
+    private _deleted = false;
```

Comment:
If I check in like this, I'll create a follow-up task to change this to tri-state: default/pendingDelete/disposed where local deletes are in the pending state until the op is ack'ed or it's rolled back.

Reply:
Opened internal bug 1029 to track cleaning this up.